src/commands/plug.rs

```diff
@@ -0,0 +1,132 @@
+    #[clap(value_name = "SOCKET_PATH", required = true)]
+    pub socket: PathBuf,
+
+    /// The path to write the output to.
+    ///
+    /// If not specified, the output will be written to stdout.
+    #[clap(long, short = 'o')]
+    pub output: Option<PathBuf>,
+}
+
+impl PlugCommand {
+    /// Executes the command.
+    pub async fn exec(&self) -> Result<()> {
+        log::debug!("executing plug command");
+        let mut graph = CompositionGraph::new();
+
+        let socket = std::fs::read(&self.socket).with_context(|| {
+            format!(
+                "failed to read socket component `{path}`",
+                path = self.socket.display()
+            )
+        })?;
+
+        let socket = Package::from_bytes("socket", None, socket, graph.types_mut())?;
+        let socket = graph.register_package(socket)?;
+        let socket_instantiation = graph.instantiate(socket);
+
+        // Plug each plug into the socket.
+        for (i, plug) in self.plugs.iter().enumerate() {
+            let plug = std::fs::read(&plug).with_context(|| {
+                format!(
+                    "failed to read plug component `{path}`",
+                    path = plug.display()
+                )
+            })?;
+            plug_into_socket(
+                &format!("plug{i}"),
+                plug,
+                socket,
+                socket_instantiation,
+                &mut graph,
+            )?;
+        }
+
+        // Check we've actually done any plugging.
+        if graph
+            .get_instantiation_arguments(socket_instantiation)
+            .next()
+            .is_none()
+        {
+            bail!("no plugs were used to plug into the socket component")
+        }
+
+        // Export all exports from the socket component.
+        for name in graph.types()[graph[socket].ty()]
+            .exports
+            .keys()
+            .cloned()
+            .collect::<Vec<_>>()
+        {
+            let export = graph.alias_instance_export(socket_instantiation, &name)?;
+            graph.export(export, &name)?;
+        }
+
+        let bytes = graph.encode(EncodeOptions::default())?;
+        match &self.output {
+            Some(path) => {
+                std::fs::write(&path, bytes).context(format!(
+                    "failed to write output file `{path}`",
+                    path = path.display()
+                ))?;
+            }
+            None => {
+                std::io::stdout()
+                    .write_all(&bytes)
+                    .context("failed to write to stdout")?;
+            }
+        }
+        Ok(())
+    }
+}
+
+/// Take the exports of the plug component and plug them into the socket component.
+fn plug_into_socket(
+    name: &str,
+    plug: Vec<u8>,
+    socket: PackageId,
+    socket_instantiation: NodeId,
+    graph: &mut CompositionGraph,
+) -> Result<(), anyhow::Error> {
+    let plug = Package::from_bytes(name, None, plug, graph.types_mut())?;
```

Comment:
Could we change this to `Package::from_file` so that the error has the context about the input path? While the context provided by `Package::from_file` doesn't mention "plug", the most relevant context is the path to the file that failed to parse.